### ISSUE TITLE
SNMP V3 never calls back if incorrect username is supplied

### ISSUE THREAD

**Issue by Steve-Mcl** [markabrahams/node-net-snmp#202]
Hi, It is more than likely I am doing something wrong but I have situation where I get no callback from the SNMP `.get` call when the `name` is not correct.

#### IMPORTNANT NOTE
If I use the correct `name`, `authProtocol` and `authKey` the lib works as expected.

#### My setup...
* This is running inside of `node-red` node `node-red-contrib-snmp` that I am upgrading to support SNMP V3 (so I dont have a simple code snippet to copy into this issue as its written around the node-red environment)
* The end device is a Netgear Sw with SNMP V3 support (an L2 switch connected directly to computer on same subnet).
* I am calling `get` with OID `1.3.6.1.2.1.1.1.0` with the following session settings...
  *  `SNMP.createV3Session(host, user, options);`
  * `host` = `192.168.x.y` 
  * `user` = `{"name":"xxxxxxxx","level":2,"authProtocol":2,"authKey":"xxxxxxxx"}`  (wrong user name deliberately provided)
  * `options` = `{"version":3,"timeout":5000,"debug":true,"transport":"udp4"}`

#### The issue

The session seems to just go round and round - below is a (quite badly written account (sorry)) but there are 2 threads in play - the timeout and the (re)sending making it difficult to write up :-/

#### Write up

Starting at `Session.prototype.send` - inside here calls `me.registerRequest (req);` ([L2285](https://github.com/markabrahams/node-net-snmp/blob/368debea76990c46eeb737fc3962d9508cf0911c/index.js#L2285))

Then in `Session.prototype.registerRequest` the `req.timer` is setup.  Eventually it times out & `req.retries` is decremented and `me.send (req);` is executed ([L2259](https://github.com/markabrahams/node-net-snmp/blob/368debea76990c46eeb737fc3962d9508cf0911c/index.js#L2259))

Then in `Session.prototype.send` inside the callback of `this.dgram.send` (LXXX) no error is received so the code them calls `me.registerRequest (req);` ([L2285](https://github.com/markabrahams/node-net-snmp/blob/368debea76990c46eeb737fc3962d9508cf0911c/index.js#L2285)).  Inside `Session.prototype.registerRequest`, the `req` object is found inside of `this.reqs` and the `setTimeout` is setup again. 

While waiting the timeout, a message arrives in `Session.prototype.onMsg`.  Once parsed, the message ([L2173](https://github.com/markabrahams/node-net-snmp/blob/368debea76990c46eeb737fc3962d9508cf0911c/index.js#L2173)) property `message.pdu.varbinds[0]` contains `{oid: '1.3.6.1.6.3.15.1.1.3.0', type: 65, value: 1475264}`.  A quick search of the internet reveals `1.3.6.1.6.3.15.1.1.3.0` to "usmStatsUnknownUserNames" "The total number of packets received by the SNMP engine which were dropped because they referenced a user that was not known to the SNMP engine." (ref: http://oid-info.com/get/1.3.6.1.6.3.15.1.1.3)

![image](https://user-images.githubusercontent.com/44235289/164323529-ac49972f-daff-4c99-9e44-8c31288a3471.png)

Stepping on though the rest of the code in `Session.prototype.onMsg`, execution ends up at line [L2215](https://github.com/markabrahams/node-net-snmp/blob/368debea76990c46eeb737fc3962d9508cf0911c/index.js#L2215) `this.sendV3Req (req.originalPdu, req.feedCb, req.responseCb, req.options, req.port, timeSyncNeeded);` and thus the cycle continues

---

Again I apologise for the truly awful write up but I have been scratching my head for some time on this & it was quite difficult to write up.

If there is something you wish me to do to help you understand the problem, please let me know.

**Comment by markabrahams** [comment]
Hi @Steve-Mcl - it seems like this particular write-up is half-way through debugging a node-red issue - have you raised this with the node-red project?

To present the problem as an actionable issue for this project, you'll need to follow the CONTRIBUTING.md guideline "Reduce the problem to a minimal reproducible example", which outlines the importance of doing some work to narrow the problem to a minimal example program. Your goal is to enable a project contributor to instantly reproduce the problem. That means I should be able to copy-paste your provided code, and get the same error/condition that you're reporting.

To help you understand the problem better, I'd suggest cutting it in half, to determine if you have a node-red problem or a device problem.

For reference, here are a set of baseline tests of the library's behaviour against a standard Net-SNMP agent, using the sample program from this library in `example/snmp-get.js` (with a different OID though - my agent didn't have the OID 1.3.6.1.2.1.1.1.0.9 from your example).  All success and error cases returned a result instantly.

With correct username, authentication algorithm and authentication key (success as expected):
```
mark@hope:~/node-net-snmp$ node ./example/snmp-get.js -v 3 -u barney -l authPriv -a sha -A correctauth -x aes -X correctpriv localhost 1.3.6.1.2.1.1.1.0
1.3.6.1.2.1.1.1.0|Linux hope 5.4.0-107-generic #121-Ubuntu SMP Thu Mar 24 16:04:27 UTC 2022 x86_64
```

With incorrect username, and correct authentication algorithm and authentication key (error as expected):
```
1.3.6.1.2.1.1.1.0|Linux hope 5.4.0-107-generic #121-Ubuntu SMP Thu Mar 24 16:04:27 UTC 2022 x86_64
mark@hope:~/node-net-snmp$ node ./example/snmp-get.js -v 3 -u INCORRECT -l authPriv -a sha -A correctauth -x aes -X correctpriv localhost 1.3.6.1.2.1.1.1.0
ResponseInvalidError: Unknown User Name
```

With correct username, incorrect authentication algorithm, and correct authentication key (error as expected):
```
mark@hope:~/node-net-snmp$ node ./example/snmp-get.js -v 3 -u barney -l authPriv -a md5 -A correctauth -x aes -X correctpriv localhost 1.3.6.1.2.1.1.1.0
ResponseInvalidError: Wrong Digest (incorrect password, community or key)
```

With correct username and authentication algorithm, and incorrect authentication key (error as expected):
```
mark@hope:~/node-net-snmp$ node ./example/snmp-get.js -v 3 -u barney -l authPriv -a sha -A INCORRECT -x aes -X correctpriv localhost 1.3.6.1.2.1.1.1.0
ResponseInvalidError: Wrong Digest (incorrect password, community or key)
```

You could run these tests with your OID 1.3.6.1.2.1.1.1.0.9 against your Netgear switch, to eliminate the node-red part of the interaction as part of your debugging effort.

**Comment by Steve-Mcl** [comment]
> it seems like this particular write-up is half-way through debugging a node-red issue - have you raised this with the node-red project?

No, it is not an issue with node-red, just that the `net-snmp` is automated via node-red. I have confirmed (as described in the OP) what values are set and sent to the `net-snmp`. 

I do have an update for you though. I have found that this issue of "no callback" does not occur when calling against `snmpd` (on Ubuntu) so may be down to a buggy v3 implementation by netgear. It still doesnt negate the point that `net-snmp` ends up going round in circles.



> You could run these tests with your OID 1.3.6.1.2.1.1.1.0.9 against your Netgear switch

![image](https://user-images.githubusercontent.com/44235289/164508506-d131c647-c2cb-4d88-b7f7-2fe8220b97bd.png)

NOTE: OID `1.3.6.1.2.1.1.1.0.9` was a deliberate error I was using while testing.  The OID I was actually using was `1.3.6.1.2.1.1.1.0`  - this is visible in the VSCode screenshot in the OP


Same tests against Ubuntu snmpd...
 
![image](https://user-images.githubusercontent.com/44235289/164509250-35d5608f-0820-49cc-acfa-f9df8654c0c8.png)

**Comment by markabrahams** [comment]
Binary chop complete!

Having established a minimal reproducible example, I'll now also need some further info, since there's an issue here with a device I don't have.

Does the same timeout occur with example command and the Netgear switch when using the `noAuthNoPriv` security level?
`node ./example/snmp-get.js -v 3 -u bad_user -l noAuthNoPriv 192.168.1.35 1.3.6.1.2.1.1.1.0`

Using the lowest security level that produces the timeout (noAuthNoPriv preferred), as per the  CONTRIBUTING.md guideline "Include packet captures" can you please post a packet capture of the timed-out interaction between the example library command and the Netgear switch?

If you could also advise on the behaviour from the Net-SNMP `snmpget` command line tool with the same Netgear switch e.g.
`snmpget -v 3 -u bad_user -l noAuthNoPriv 192.168.1.35 1.3.6.1.2.1.1.1.0
`
and include a packet capture of this as well, that would be useful info as a point of comparison.

**Comment by Steve-Mcl** [comment]
> Does the same timeout occur with example command and the Netgear switch when using the noAuthNoPriv security level?
node ./example/snmp-get.js -v 3 -u bad_user -l noAuthNoPriv 192.168.1.35 1.3.6.1.2.1.1.1.0

Yes - the "no callback" issue persists...
![image](https://user-images.githubusercontent.com/44235289/164767333-0399154b-6106-4ffc-aaa9-6c2693756750.png)

> If you could also advise on the behaviour from the Net-SNMP snmpget command line tool with the same Netgear switch e.g.
snmpget -v 3 -u bad_user -l noAuthNoPriv 192.168.1.35 1.3.6.1.2.1.1.1.0

Using snmpget works as expected...
![image](https://user-images.githubusercontent.com/44235289/164767627-e823258b-ed57-4645-8313-437df0915180.png)



[snmpget-vs-netsnmp.zip](https://github.com/markabrahams/node-net-snmp/files/8543006/snmpget-vs-netsnmp.zip)

**Comment by markabrahams** [comment]
Thanks heaps for the info @Steve-Mcl, particularly the packet captures - these reveal where the problem lies.

In summary, the check for allowing a second report PDU for SNMPv3 time synchronisation was for engineBoots OR engineUptime to be set to zero.  In the case of the Netgear, engineBoots was zero, but engineUptime was non-zero.  I believe just one of these needs to be non-zero to suggest time synchronisation is not required, and therefore a second report PDU implies an error condition that should be supplied to the caller.

I've created pull request #203 as a prospective fix for this, but would like this reviewed first by the author of this feature, @gavinaiken, if possible.

@Steve-Mcl, if you're able to test the fix by installing the `time-sync-fix` branch and re-testing, e.g. by changing the package.json "net-snmp" line to:
  "net-snmp": "markabrahams/node-net-snmp#time-sync-fix"

and re-running `npm install` that would also be much appreciated.

**Comment by markabrahams** [comment]
@gavinaiken - if you're available to review #203, that would be much appreciated.

The original time sync requirement check introduced in commit
https://github.com/markabrahams/node-net-snmp/pull/66/commits/deef3732b04e0ed45967453aa1430660b25f9662

was:
`! message.msgSecurityParameters.msgAuthoritativeEngineBoots && ! message.msgSecurityParameters.msgAuthoritativeEngineTime`

which I believe was the correct logic.

In commit
https://github.com/markabrahams/node-net-snmp/pull/71/commits/420a9c7c630d20c45352aaf3f3ff445c3dd6e088

you changed this to:
`! message.msgSecurityParameters.msgAuthoritativeEngineBoots || ! message.msgSecurityParameters.msgAuthoritativeEngineTime`

which I think is flawed logic, as some devices (e.g. the Netgear switch that @Steve-Mcl mentions above) genuinely have zero engine boots and send this through for time sync (along with a non-zero uptime)

I just wanted to double-check that there wasn't some other reason behind the switch from AND to OR here that I'm missing, before I merge a #203 to revert back to the AND logic, which should fix this problem.

Also, if you could confirm the code in PR #203 works OK for the 2-report time sync device if you still have one available, that would confirm a change in logic isn't going to break that scenario (I don't think it should, but better safe than sorry!)

Thanks!

**Comment by Steve-Mcl** [comment]
@markabrahams I can confirm this now works as expected...

![image](https://user-images.githubusercontent.com/44235289/164896916-53254c41-a07a-4aa3-aba5-8684602bb8a1.png)


Any idea on the timing of an update to NPM please?

**Comment by gavinaiken** [comment]
Seems to make sense to me, but I don't think I would have made that change unless it seemed to be necessary for the Cisco device I was working with at the time. I should be able to test this change against the same devices on Monday morning if you don't mind waiting until then.

**Comment by markabrahams** [comment]
@gavinaiken - yeah that's exactly what I was thinking - it would be unusual to make a logic change like that without some sort of reasoning - so I was just cautious in reverting it without understanding the background, making sure there wasn't some other aspect / side-effect of this I hadn't considered.

@Steve-Mcl - if all checks out with testing from @gavinaiken - publishing the fix in the npm shouldn't be more than a few days away.  If there are issues with regressions, that could extend the timeline.

**Comment by gavinaiken** [comment]
I have tested this change today on the 2 Cisco devices which were giving me trouble originally and they are fine with this change. I also tested going back to version 2.5.6 of the library, ie before my change, to verify that I would still see the `ResponseInvalidError: Unexpected Report PDU` error, which I could, so I believe this change should be good. Not sure why I made it but am guessing it was an early stab at fixing the issue which didn't pan out and then I didn't realise it wasn't necessary along with the other changes which were.

In summary :shipit: 👍
